### PR TITLE
server: adds batch RPC request and method counters

### LINKED ISSUE #65753: kvserver: add metrics for rpcs received by the server
**Is your feature request related to a problem? Please describe.**

We should have request counters for the server side. We have them for the client side. 

See `distsender.rpc..*` and https://github.com/cockroachdb/cockroach/blob/83a204d93f94b332253c937da7ee271e214b4860/pkg/kv/kvclient/kvcoord/dist_sender.go#L802-L804

Epic: CRDB-8500

### PR DESCRIPTION
**Pull request by jtsiros**
server: adds batch RPC request and method counters
    
 Previously, we were only tracking RPC batch requests client-side.
 This change adds server-side batch RPC count and batch RPC method
 count metrics with the following prefixes:
    
 rpc.batches.recv
 rpc.method.*.recv
    
Release note (api change): added new batch RPC and batch method 
counters now visible in DB console and _status/vars.

closes #65753.

### CODE REVIEW

**Review by knz — COMMENTED** [review]
I'm missing some unit tests?

> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72767#-:-Mo_A4rd1nNtHIDSl0I7:bjdc7i6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @rauchenstein)

**Review by jtsiros — COMMENTED** [review]
Ack - I was in the process of starting a discussion here to make sure I added these metrics in the right place. I was going to add tests  after you and Jay looked it over.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72767#-:-Mo_Cz-T-CNwTQ-wttwS:b-u4fnno)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @rauchenstein)

File: pkg/server/node.go [hunk 1]
```diff
@@ -114,6 +115,20 @@ This metric is thus not an indicator of KV health.`,
...
+		Help:        "Number of %s requests processed",
+		Measurement: "RPCs",
+		Unit:        metric.Unit_COUNT,
+	}
+
+	metaInternalBatchRPCCount = metric.Metadata{
+		Name:        "internal.rpc.batches",
+		Help:        "Number of batches processed",
+		Measurement: "Batches",
+		Unit:        metric.Unit_COUNT,
```
Comment (erikgrinaker) [inline pkg/server/node.go:130]: What's the motivation for the `internal.` prefix? Seems like just using a `rpc.` prefix would suffice? In that case we should put the method counts under a separate group though, something like `rpc.method.%s.recv` or `rpc.request.%.recv`.

Aside from the bikeshedding, LGTM -- thanks for adding these!

Comment (jtsiros) [reply pkg/server/node.go:130]: I originally had `rpc.` but added `internal.` thinking that adding the service would be more clear. Now that you mention it, it seems redundant. Thanks for the feedback!

**Review by knz — COMMENTED** [review]
Sorry for the late review.
For the KV batch requests, this PR is a-OK.

However, it only tracks requests to the `Batch` RPC endpoint. We have dozens of other RPCs, and many of those are worth counters as well.

Instead of an opt-in call to a metric function inside each handler method, I wonder how much leverage we'd have by adding a counter increase  as an additional interceptor, or an extension of the existing interceptor, in `newGRPCServer()` (in `pkg/server/grpc_server.go`) - this receives the RPC method name as a string argument, which we can use as key in a map of metrics.

> Reviewed 2 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72767#-:-MpfIKlT9BaLSYgniGml:b-4llwhz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker, @jtsiros, @knz, and @rauchenstein)

___
*[pkg/server/node.go, line 976 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/72767#-MpfIINEABx4hCE375FP:-MpfIINEABx4hCE375FQ:b-25z16g) ([raw file](https://github.com/cockroachdb/cockroach/blob/447b4e2813ebc291d6301253b5aa2d5918aaafd0/pkg/server/node.go#L976)):*
> ```Go
> // request methods.
> func (n *Node) incrementBatchCounters(ba *roachpb.BatchRequest) {
> 	n.metrics.BatchCount.Inc(1)
> ```

this is a problem (see my comment at top)
___
*[pkg/server/node.go, line 977 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/72767#-MpfIHHr7BCUilLn7QtG:-MpfIHHt-tEOMMVHaucI:beb8v9x) ([raw file](https://github.com/cockroachdb/cockroach/blob/447b4e2813ebc291d6301253b5aa2d5918aaafd0/pkg/server/node.go#L977)):*
> ```Go
> func (n *Node) incrementBatchCounters(ba *roachpb.BatchRequest) {
> 	n.metrics.BatchCount.Inc(1)
> 	for _, ru := range ba.Requests {
> ```

this is OK

**Comment by jtsiros** [comment]
___
*[pkg/server/node.go, line 130 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72767#-MoZtM0LsF52cmYbD-P4-r2-130:-MpgwgeB3Cp_1k6ASLFh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/007639caa2cf85a0968701d0be7b4d8e7f2f27cc/pkg/server/node.go#L130)):*
<details><summary><i>Previously, jtsiros (Jon Tsiros) wrote…</i></summary><blockquote>

I originally had `rpc.` but added `internal.` thinking that adding the service would be more clear. Now that you mention it, it seems redundant. Thanks for the feedback!
</blockquote></details>

Done.

**Review by rauchenstein — COMMENTED** [review]
> Reviewed 1 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72767)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @jtsiros)

**Review by jtsiros — COMMENTED** [review]
Thanks. I asked the same question [here](https://cockroachlabs.slack.com/archives/C0KB9Q03D/p1635265818033100) and this was intended for `Batch` RPCs only. I do see the value in doing this within the gRPC interceptor rather than having to opt in, but I think it's out of scope for this PR. We can always ship this change and refactor at a later time as this will require a more thorough refactor in other places we gather metrics too.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72767#-:-MproDP3BEzU-A8Ymn2E:b362cvw)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker, @jtsiros, and @knz)

**Comment by knz** [comment]
ok then
can you file a followup issue for the more general problem, then this is good to go

**Comment by jtsiros** [comment]
bors r=knz,erikgrinaker

**Comment by thtruo** [comment]
Hi @jtsiros I was grooming the DOCS backlog with @taroface and saw a corresponding docs issue here https://github.com/cockroachdb/docs/issues/12497

Just wanted to double check, does your change mean you've added 2 new Prometheus metrics? And if so, are they currently displayed in any of our metrics pages?

**Comment by jtsiros** [comment]
@thtruo, there is one new prometheus metric `rpc_batches_recv` that counts batch rpcs and then 47 different batch requests metrics (see [L289](https://github.com/cockroachdb/cockroach/pull/72767/files#diff-a7401d803ec09fc55628cdfc3f988221338af161a44fb60f2481efbaef9115d1#L289)). I don't think we are exposing these as graphs in the console.

**Pull request merged**